src/lucky_router/match_finder.cr

```diff
@@ -20,20 +21,19 @@ class LuckyRouter::MatchFinder(T)
   # This uses the magic/pain of recursion to continue matching fragments
   # until there is a match in the final fragment, otherwise it returns `NoMatch`
   def run : Match(T) | NoMatch
-    add_to_params if has_param?
+    until parts.empty?
```

Comment:
I'm curious... I did a little benchmark on what performs better between using `loop` and using `until`, and `loop` is a "little" faster.

```crystal
require "benchmark"
PARTS = (0...1_000_000).to_a
def with_until
  local = PARTS
  until local.empty?
    1 + 1
    local.shift
  end
end

def with_loop
  local = PARTS
  loop do
    break if local.empty?
    1 + 1
    local.shift
  end
end

Benchmark.ips do |x|
  x.report("with_until") { with_until }
  x.report("with_loop") { with_loop }
end
```

```
[18:50PM] sandbox$ ./a
with_until  24.68M ( 40.53ns) (± 2.74%)  0.0B/op  11.22× slower
 with_loop 276.92M (  3.61ns) (± 1.49%)  0.0B/op        fastest
```

From your perspective, if we used `loop` instead, would that throw anything off? Change implementation or anything?

```crystal
loop do
  break if parts.empty?

  # ... other stuff
  parts.shift
end
```

Reply:
actually.. the break could go after the shift to do 1 less loop setup... 🤔

Reply:
It seems like switching to loop made it 0.2ms faster which is not a huge difference. The smaller difference between your code and mine is that you are running your loops for much longer than the router's paths. I'd rather switch to a for...each loop or a reduce call in future changes. It might be slower but I find the loops ands and mutations going on here to make the code unclear.